# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [1]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv')
pd.set_option('display.max_columns', None)
df.head()

,Unnamed: 0,Customer,State,Customer Lifetime Value,Response,Coverage,Education,Effective To Date,EmploymentStatus,Gender,Income,Location Code,Marital Status,Monthly Premium Auto,Months Since Last Claim,Months Since Policy Inception,Number of Open Complaints,Number of Policies,Policy Type,Policy,Renew Offer Type,Sales Channel,Total Claim Amount,Vehicle Class,Vehicle Size,Vehicle Type
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2/18/11,Employed,M,48029,Suburban,Married,61,7.0,52,0.0,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,NaN
1,1,KX64629,California,2228.525238,No,Basic,College,1/18/11,Unemployed,F,0,Suburban,Single,64,3.0,26,0.0,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,NaN
2,2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2/10/11,Employed,M,22139,Suburban,Single,100,34.0,31,0.0,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,1/11/11,Employed,M,49078,Suburban,Single,97,10.0,3,0.0,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A
4,4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,1/17/11,Medical Leave,F,23675,Suburban,Married,117,NaN,31,NaN,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,NaN


In [2]:
df.shape

(10910, 26)

In [3]:
df.info() #para tener visibilidad del nombre de las columnas. 

<class 'pandas.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     10910 non-null  int64  
 1   Customer                       10910 non-null  str    
 2   State                          10279 non-null  str    
 3   Customer Lifetime Value        10910 non-null  float64
 4   Response                       10279 non-null  str    
 5   Coverage                       10910 non-null  str    
 6   Education                      10910 non-null  str    
 7   Effective To Date              10910 non-null  str    
 8   EmploymentStatus               10910 non-null  str    
 9   Gender                         10910 non-null  str    
 10  Income                         10910 non-null  int64  
 11  Location Code                  10910 non-null  str    
 12  Marital Status                 10910 non-null  str    
 1

In [4]:
df.columns = df.columns.str.lower().str.replace(' ', '_')   #vamos a corregir el nombre de las columnas para empezar con el analisis. 
df.head()

,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,income,location_code,marital_status,monthly_premium_auto,months_since_last_claim,months_since_policy_inception,number_of_open_complaints,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2/18/11,Employed,M,48029,Suburban,Married,61,7.0,52,0.0,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,NaN
1,1,KX64629,California,2228.525238,No,Basic,College,1/18/11,Unemployed,F,0,Suburban,Single,64,3.0,26,0.0,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,NaN
2,2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2/10/11,Employed,M,22139,Suburban,Single,100,34.0,31,0.0,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,1/11/11,Employed,M,49078,Suburban,Single,97,10.0,3,0.0,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A
4,4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,1/17/11,Medical Leave,F,23675,Suburban,Married,117,NaN,31,NaN,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,NaN


In [5]:
#1. Crear un dataframe con clientes que: tiene pocas reclamaciones (inferior a 1000) y han respondido Si a la campaña. Lo que tenemos que hacer es filtrar los datos. 

low_claim = df['total_claim_amount'] < 1000
yes_response = df['response'] == 'Yes'

filtered_df = df[low_claim & yes_response]

filtered_df.head()


,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,income,location_code,marital_status,monthly_premium_auto,months_since_last_claim,months_since_policy_inception,number_of_open_complaints,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,1/11/11,Employed,M,49078,Suburban,Single,97,10.0,3,0.0,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A
8,8,FM55990,California,5989.773931,Yes,Premium,College,1/19/11,Employed,M,66839,Suburban,Single,154,33.0,24,0.0,1,Personal Auto,Personal L1,Offer2,Branch,739.200000,Sports Car,Medsize,NaN
15,15,CW49887,California,4626.801093,Yes,Basic,Master,1/16/11,Employed,F,79487,Suburban,Divorced,114,20.0,87,0.0,1,Special Auto,Special L1,Offer2,Branch,547.200000,SUV,Medsize,NaN
19,19,NJ54277,California,3746.751625,Yes,Extended,College,2/26/11,Employed,F,41479,Rural,Married,94,14.0,38,1.0,1,Personal Auto,Personal L2,Offer2,Call Center,19.575683,Two-Door Car,Large,A
27,27,MQ68407,Oregon,4376.363592,Yes,Premium,Bachelor,2/28/11,Employed,F,63774,Rural,Divorced,111,18.0,63,0.0,1,Personal Auto,Personal L3,Offer2,Agent,60.036683,Four-Door Car,Medsize,NaN


In [6]:
print("Number of filtered rows:", filtered_df.shape[0])
filtered_df[['total_claim_amount', 'response']].head()

Number of filtered rows: 1399


,total_claim_amount,response
3,484.013411,Yes
8,739.200000,Yes
15,547.200000,Yes
19,19.575683,Yes
27,60.036683,Yes


In [7]:
#2. Calcular medias de monthly_premium y de customer_lifetime_value agrupado por: policy_type y gender considerando la response: yes y trabajando con el original dataframe y compararlo con total_claim_amount

yes_customers = df[df['response'] == 'Yes']

profitability_analysis = yes_customers.groupby(
    ['policy_type', 'gender']
).agg({
    'customer_lifetime_value': 'mean',
    'monthly_premium_auto': 'mean',
    'total_claim_amount': 'mean'
}).reset_index()

profitability_analysis


,policy_type,gender,customer_lifetime_value,monthly_premium_auto,total_claim_amount
0,Corporate Auto,F,7712.628736,94.301775,433.738499
1,Corporate Auto,M,7944.465414,92.188312,408.582459
2,Personal Auto,F,8339.791842,98.998148,452.965929
3,Personal Auto,M,7448.383281,91.085821,457.010178
4,Special Auto,F,7691.584111,92.314286,453.280164
5,Special Auto,M,8247.088702,86.343750,429.527942


In [8]:
profitability_analysis.sort_values(
    by='customer_lifetime_value',
    ascending=False
)

,policy_type,gender,customer_lifetime_value,monthly_premium_auto,total_claim_amount
2,Personal Auto,F,8339.791842,98.998148,452.965929
5,Special Auto,M,8247.088702,86.343750,429.527942
1,Corporate Auto,M,7944.465414,92.188312,408.582459
0,Corporate Auto,F,7712.628736,94.301775,433.738499
4,Special Auto,F,7691.584111,92.314286,453.280164
3,Personal Auto,M,7448.383281,91.085821,457.010178


Insights del analisis: 
- Las mujeres con pólizas Personal Auto presentan el mayor Customer Lifetime Value promedio (8.340) y también el mayor Monthly Premium Auto (99), lo que las convierte en el segmento potencialmente más rentable.
- Los hombres con pólizas Special Auto también representan clientes de alto valor, con un Customer Lifetime Value promedio de 8.247 y un Total Claim Amount relativamente moderado (430).
- Los hombres con pólizas Personal Auto registran el mayor Total Claim Amount promedio (457), lo que indica un mayor nivel de riesgo para la compañía.
- Los clientes de Corporate Auto muestran niveles de reclamaciones más bajos (409–434), aunque también generan un Customer Lifetime Value inferior al de los segmentos Personal Auto y Special Auto.
- En conjunto, el segmento Personal Auto - Female parece ofrecer el mejor equilibrio entre rentabilidad y riesgo, mientras que Personal Auto - Male destaca por su rentabilidad pero también por presentar el mayor coste asociado a reclamaciones.

In [9]:
#3. Analizar cuantos clientes hay por state que tienen poliza. Y filtra el resultado incluyendo solo aquellos states con mas de 500 paises. 

state_customers = (
    df.groupby('state')['customer']
    .count()
    .reset_index(name='number_of_customers')
)

state_customers

,state,number_of_customers
0,Arizona,1937
1,California,3552
2,Nevada,993
3,Oregon,2909
4,Washington,888


In [10]:
filtered_states = state_customers[
    state_customers['number_of_customers'] > 500
]

filtered_states

,state,number_of_customers
0,Arizona,1937
1,California,3552
2,Nevada,993
3,Oregon,2909
4,Washington,888


In [11]:
print("Number of states with more than 500 customers:", filtered_states.shape[0])

Number of states with more than 500 customers: 5


The analysis shows the number of customers in each state. After applying the filter, only states with more than 500 customers are retained.

This helps identify the states where the company has the largest customer base and could focus its marketing and retention strategies.

In [12]:
#4. Calcula el max, min, mediana de customer lifetime value por education level y gender. 
clv_stats = df.groupby(
    ['education', 'gender']
)['customer_lifetime_value'].agg(
    ['max', 'min', 'median']
).reset_index()

clv_stats


,education,gender,max,min,median
0,Bachelor,F,73225.95652,1904.000852,5640.505303
1,Bachelor,M,67907.27050,1898.007675,5548.031892
2,College,F,61850.18803,1898.683686,5623.611187
3,College,M,61134.68307,1918.119700,6005.847375
4,Doctor,F,44856.11397,2395.570000,5332.462694
5,Doctor,M,32677.34284,2267.604038,5577.669457
6,High School or Below,F,55277.44589,2144.921535,6039.553187
7,High School or Below,M,83325.38119,1940.981221,6286.731006
8,Master,F,51016.06704,2417.777032,5729.855012
9,Master,M,50568.25912,2272.307310,5579.099207


In [13]:
clv_stats.sort_values(
    by='median',
    ascending=False
)

,education,gender,max,min,median
7,High School or Below,M,83325.38119,1940.981221,6286.731006
6,High School or Below,F,55277.44589,2144.921535,6039.553187
3,College,M,61134.68307,1918.119700,6005.847375
8,Master,F,51016.06704,2417.777032,5729.855012
0,Bachelor,F,73225.95652,1904.000852,5640.505303
2,College,F,61850.18803,1898.683686,5623.611187
9,Master,M,50568.25912,2272.307310,5579.099207
5,Doctor,M,32677.34284,2267.604038,5577.669457
1,Bachelor,M,67907.27050,1898.007675,5548.031892
4,Doctor,F,44856.11397,2395.570000,5332.462694


Insights:
- Hay una gran diferencia entre el valor máximo y la mediana, lo que indica pocos clientes altamente valiosos. 
- Los clientes con menor nivel educativo pueden alcanzar valores de vida muy altos, lo que rompe la idea de que mayor educación implica mayor valor.
- El género no tiene un impacto consistente en el valor del cliente, ya que su efecto varía según el nivel educativo. 
